### ISSUE TITLE
[BUG]Nan in gradients of scaled_dot_product_attention operation with mem_efficient backend

### ISSUE THREAD

**Issue by walkacross** [pytorch/pytorch#125674]
### 🐛 Describe the bug

#  reproduce the bug
@mstebelev I found out that memory efficient attention kernel on float32 cuda tensors gives nan gradients despite inputs and incoming gradient are reasonably limited. Math backend doesn't produce nans with this input.

below for a minimal and simple script to reproduce the issue(from @otto-dev)
~~~
$ pip show torch
Name: torch
Version: 2.4.0.dev20240427+cu121

$ python --version
Python 3.11.6
~~~

~~~
import torch
import torch.nn as nn
import torch.optim as optim

embed_dim = 1024
batch_size = 32
seq_length = 50
num_iterations = 100000
learning_rate = 0.01
device = torch.device("cuda")
torch.autograd.set_detect_anomaly(True)

# Initialize model
model = nn.TransformerEncoderLayer( embed_dim, 16, dropout=0.01, batch_first=True).to(device)
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
inputs = torch.full((seq_length, batch_size, embed_dim), 1000.0, device=device)

# Training loop
for i in range(num_iterations):
    print(f'Iteration {i + 1}')
    optimizer.zero_grad()
    output = model(inputs)
    loss = output.mean()
    loss.backward()
    optimizer.step()
~~~

Turns out it's very easy to reproduce: Train a TransformerEncoderLayer on a tensor that contains nothing but the number 1000 a few times. The higher the number (here, 1000), the quicker the error occurs.

Reproduction script:
~~~
$ python repro.py 
Iteration 1
Iteration 2
Iteration 3
Iteration 4
Iteration 5
/otto-dev/venv/lib/python3.11/site-packages/torch/autograd/graph.py:767: UserWarning: Error detected in ScaledDotProductEfficientAttentionBackward0. Traceback of forward call that caused the error:
  File "/otto-dev/repro.py", line 22, in <module>
    output = model(inputs)
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/modules/transformer.py", line 748, in forward
    x = self.norm1(x + self._sa_block(x, src_mask, src_key_padding_mask, is_causal=is_causal))
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/modules/transformer.py", line 756, in _sa_block
    x = self.self_attn(x, x, x,
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/modules/activation.py", line 1266, in forward
    attn_output, attn_output_weights = F.multi_head_attention_forward(
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/nn/functional.py", line 5527, in multi_head_attention_forward
    attn_output = scaled_dot_product_attention(q, k, v, attn_mask, dropout_p, is_causal)
 (Triggered internally at ../torch/csrc/autograd/python_anomaly_mode.cpp:111.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
Traceback (most recent call last):
  File "/otto-dev/repro.py", line 24, in <module>
    loss.backward()
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/_tensor.py", line 523, in backward
    torch.autograd.backward(
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/autograd/__init__.py", line 267, in backward
    _engine_run_backward(
  File "/otto-dev/venv/lib/python3.11/site-packages/torch/autograd/graph.py", line 767, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Function 'ScaledDotProductEfficientAttentionBackward0' returned nan values in its 0th output.
~~~
This exception will not be triggered immediately during training, but will be triggered after training for many epochs.

# the issue from memory efficient attention kernel and Math backend doesn't produce nans with this input. 

code from @SamAdamDay 
~~~
import torch
import torch.nn as nn
import torch.optim as optim
from torch.nn.attention import SDPBackend, sdpa_kernel

embed_dim = 1024
batch_size = 32
seq_length = 50
num_iterations = 1000
learning_rate = 0.01
device = torch.device("cuda")
torch.autograd.set_detect_anomaly(True)

# Initialize model
model = nn.TransformerEncoderLayer( embed_dim, 16, dropout=0.01, batch_first=True).to(device)
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
inputs = torch.full((seq_length, batch_size, embed_dim), 1000.0, device=device)

# Training loop
with sdpa_kernel(SDPBackend.MATH):
    for i in range(num_iterations):
        print(f'Iteration {i + 1}')
        optimizer.zero_grad()
        output = model(inputs)
        loss = output.mean()
        loss.backward()
        optimizer.step()
~~~

# more evidence and disscussion
* https://github.com/pytorch/pytorch/issues/119320
* https://discuss.pytorch.org/t/function-scaled-dot-product-efficient-attention-backward0-returned-nan-values-in-its-0th-output/191752

**It is clear that this is a very serious bug that can affect many researcher's model train process and experimental results. It should be fixed quickly**

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @bhosmer @cpuhrsch @erichan1 @drisspg

### Versions

2.1.0+cu121 - 2.4.0.dev20240427+cu121

**Comment by drisspg** [comment]
One Update
I tried locally compiling with `string(APPEND CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -fmad=false")`
And the error still repros


Also while using: [NanInfdetect](https://github.com/drisspg/transformer_nuggets/blob/522999a840d665a9bb23d96a955f54e84c2a2b22/transformer_nuggets/utils/tracing.py#L121)

I was able to determine that this training loops is encouraging very large values  for q, k , v
``` Shell
4.9956e+08, -4.9956e+08, -4.9956e+08,  ..., -4.9956e+08,
          -4.9956e+08, -4.9956e+08],
         [ 2.7200e+08,  2.7200e+08,  2.7200e+08,  ...,  2.7200e+08,
           2.7200e+08,  2.7200e+08],
           
```

If you set `norm_first = True` I am unable to reproduce the error

**Comment by Glexin** [comment]
I met the same problem, when I use cpu backend, it's ok, nan grad appears when use gpu. I debug and find out inf grad to q, k, v appears through "attn_output = scaled_dot_product_attention(q, k, v, attn_mask, dropout_p, is_causal)" this line in torch/nn/functional.py, I can't trace deeper when it goes into C.

**Comment by otto-dev** [comment]
TL;DR: Somebody may have identified what the issue is here, something to do with `dq, dk, dv` not being overwritten properly in the backwards pass - please have a look :slightly_smiling_face: https://github.com/Dao-AILab/flash-attention/issues/41#issuecomment-1303003442 @walkacross @drisspg @SamAdamDay 

I'm the author of the reproduction script above. I observed the exact same NaN behavior when using https://github.com/Dao-AILab/flash-attention as a drop-in implementation for `TransformerEncoderLayer` instead of using `pytorch's` implementation. That is, after a non-deterministic number of iterations there will be NaNs, with a `FlashAttnQKVPackedFuncBackward` error referenced in the issue below.

This led me to an issue in the above repository where somebody apparently correctly identifies the cause of the issue: https://github.com/Dao-AILab/flash-attention/issues/41#issuecomment-1303003442

Interestingly, in both implementations the error is heavily influenced by the size of `embed_dim`. Changing `embed_dim` from 128 to 64 makes the error go away using `Dao-AILab/flash-attention`, while it takes thousands of iterations to trigger the error using `pytorch`. Changing `embed_dim` to 32 makes the error go away in both implementations.

Unfortunately my transformer-foo and lingo is not educated enough to verify the issue and solution. All I can see is that the behavior is the same.

For reference, https://github.com/Dao-AILab/flash-attention/ reproduction script:

```py
import torch
import torch.optim as optim
import torch.nn.functional as F
from functools import partial
from flash_attn.modules.block import Block
from flash_attn.modules.mha import MHA
from flash_attn.modules.mlp import Mlp

embed_dim = 256 # !!! change this to 64 and the error will not be observable !!!
batch_size = 16
num_heads = 8
seq_length = 512
dim_feedforward = 1024
learning_rate = 0.01
device = torch.device("cuda")
torch.set_default_dtype(torch.bfloat16)
torch.autograd.set_detect_anomaly(True)

# Initialize model
model = Block( # TransformerEncoderLayer
        embed_dim,
        mixer_cls=partial(
            MHA,
            num_heads=num_heads,
            use_flash_attn=True,
            rotary_emb_dim=0,
        ),
        mlp_cls=partial(Mlp, hidden_features=dim_feedforward),
        resid_dropout1=0.0,
        resid_dropout2=0.0,
        prenorm=False,
    ).to(device)

optimizer = optim.Adam(model.parameters(), lr=learning_rate)
inputs = torch.full((batch_size, seq_length, embed_dim), 1000.0, device=device)

# Training loop
for i in range(999999):
    print(f'Iteration {i + 1}')
    optimizer.zero_grad()
    output = model(inputs)
    loss = output.mean()
    loss.backward()
    optimizer.step()
```

Error:

```sh
RuntimeError: Function 'NativeLayerNormBackward0' returned nan values in its 0th output.
```

updated `pytorch` repro script (safer numbers):
```py
import torch
import torch.nn as nn
import torch.optim as optim

embed_dim = 128 # !!! change this to 64 and the error will occur only after thousands of iterations !!!
batch_size = 512
seq_length = 16
nhead = 8
num_iterations = 100000
learning_rate = 0.01
device = torch.device("cuda")
torch.autograd.set_detect_anomaly(True)

# Initialize model
model = nn.TransformerEncoderLayer(embed_dim, nhead=nhead, dropout=0.0, batch_first=True, dim_feedforward=1024).to(device)
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
inputs = torch.full((batch_size, seq_length, embed_dim), 1000.0, device=device)

# Training loop
for i in range(num_iterations):
    print(f'Iteration {i + 1}')
    optimizer.zero_grad()
    output = model(inputs)
    loss = output.mean()
    loss.backward()
    optimizer.step()
```

**Comment by Oktai15** [comment]
Any updates according to this important problem?

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @bhosmer @cpuhrsch @erichan1 @drisspg @ptrblck

**Comment by YvanYin** [comment]
Does someone know how to avoid these NaN issues?

**Comment by sjoshi804** [comment]
Following up on this! Any updates?

**Comment by MrForExample** [comment]
Still same issue in pytorch 2.8.0

**Comment by drisspg** [comment]
Sorry it took quite a long time to get to this, but for the specific case above I will explain why this is an edgecase -> why we dont plan to change this and why im closing this issue. The TLDR though is that for these insanely large inputs this is a property of FA/safe softmax's backward implementation. If you need a fused implementation that gets you further along you can try flex_attention with tf32 disabled* and this will likely push off the nan further away but the delta in grads is still quite larger.

### Details
The backward of fa does not recompute the denom of softmax in the backward. And instead it onlys aves the LSE and uses the fact that we can recover the normalized scores via 

Recover the normalized scores via

$$
\begin{aligned}
\ell_i &= s_i - \mathrm{LSE} \\
p_i &= \exp(s_i - \mathrm{LSE}) \\
&= \frac{\exp(s_i)}{\exp(\mathrm{LSE})} \\
&= \frac{\exp(s_i)}{\sum_j \exp(s_j)} \\
&= \mathrm{softmax}(s)_i
\end{aligned}
$$

We are dealing with exponentials so small nums get big quickly. In this case for float32 

<img width="236" height="169" alt="Image" src="https://github.com/user-attachments/assets/c7f640c0-8473-4fb6-8f0a-ad6a9108f018" />

It does not take much to saturate. But this is softmax so we shouldnt ever get a number > 0 because how can the denom be > than the numerator.

Typically we are dealing with nice scores that are well within fp32's range,  this means that a ulp here or a ulp there doesnt acutally amount to much rounding.  But what if your inputs are ginormous. Well then one ULP wobble the wrong way can amount to potentially a smaller LSE or a larger SCORE than expected.

This leads us to the pathological shape that I think this issue is hitting. 

Assume all your scores are about equal and that they are very large and have a value of A. 
`softmax(A, A, ..., A) = exp(A) / sum_j exp(A) = exp(A) / (N * exp(A)) = 1/N`
 which means LSE = A + log(N).

 In the repro, seq_length = 16, so N = 16 and log(N) ≈ 2.77.

 For me around iteration 200, scores reach ≈ 1.34e10. At that magnitude, one fp32 ULP is 1024 and A + log(N)  will (ideally) get rounded to A becuase log(N) < 1 ulp. That means even before any rounding we already cant get back what we need from the LSE and the info is lost.  So in the best case where score == LSE exactly, we’d get exp(0) = 1 for every entry reconstructed P_i = 1 instead of the true 1/16. The probabilities are wrong (don’t sum to 1), but at least they’re finite. But any forward/backward GEMM rounding wobble that pushes the recomputed `score` above the saved LSE by even one ULP gives `score - LSE = +1024`, and exp(...) overflows in the backward reconstruction → NaNs in dq/dk/dv.

We can add a full fp32 support to mem-effiecnet attention but I would like that to be a separate issue.
